### ISSUE TITLE
minor issue in XMP format sidecar - dc:description and dc:title

### ISSUE THREAD

**Issue by narensankar0529** [RhetTbull/osxphotos#350]
when description and title are empty - exiftool does not like how osxphotos writes the tags.

And it appears that Windows Explorer and some apps don't like empty description and title as osxphotos outputs it. Preferred way is how Exiftool writes them - 

So instead of 

<dc:description></dc:description>
        <dc:title></dc:title>

Exiftool prefers they look as follows -

<dc:description>
   <rdf:Alt>
    <rdf:li xml:lang='x-default'/>
   </rdf:Alt>
  </dc:description>

  <dc:title>
   <rdf:Alt>
    <rdf:li xml:lang='x-default'/>
   </rdf:Alt>
  </dc:title>

**Comment by RhetTbull** [comment]
Thanks.  I'll take a look this weekend.

Here's an exiftool generated XMP file:

```xml
<?xpacket begin='﻿' id='W5M0MpCehiHzreSzNTczkc9d'?>
<x:xmpmeta xmlns:x='adobe:ns:meta/' x:xmptk='Image::ExifTool 12.00'>
<rdf:RDF xmlns:rdf='http://www.w3.org/1999/02/22-rdf-syntax-ns#'>

 <rdf:Description rdf:about=''
  xmlns:dc='http://purl.org/dc/elements/1.1/'>
  <dc:description>
   <rdf:Alt>
    <rdf:li xml:lang='x-default'>My description</rdf:li>
   </rdf:Alt>
  </dc:description>
  <dc:title>
   <rdf:Alt>
    <rdf:li xml:lang='x-default'>Title Here</rdf:li>
   </rdf:Alt>
  </dc:title>
 </rdf:Description>

 <rdf:Description rdf:about=''
  xmlns:photoshop='http://ns.adobe.com/photoshop/1.0/'>
  <photoshop:DateCreated>2020-09-19T14:36:26.719</photoshop:DateCreated>
 </rdf:Description>

 <rdf:Description rdf:about=''
  xmlns:xmp='http://ns.adobe.com/xap/1.0/'>
  <xmp:CreateDate>2020-09-19T14:36:26.719</xmp:CreateDate>
  <xmp:CreatorTool>13.7</xmp:CreatorTool>
  <xmp:ModifyDate>2020-09-19T14:36:26</xmp:ModifyDate>
 </rdf:Description>
</rdf:RDF>
</x:xmpmeta>
<?xpacket end='w'?>
```

I do have tests that use exiftool to validate the XMP and json files but exiftool says the file is valid--but when trying to apply tags, it says it's not:

```
exiftool -xmp=IMG_3092.heic.xmp IMG_3092.heic
Warning: Invalid XMP data for XMP:XMP
    0 image files updated
    1 image files unchanged

exiftool -validate IMG_3092.heic.xmp
Validate                        : OK
```

It would be really helpful to have a tool that validates an XMP file but I've been unable to find such.

**Comment by RhetTbull** [comment]
What error are you getting from exiftool?  Here's an output from an osxphotos generated XMP file and exiftool seems to accept it:

```
exiftool -tagsfromfile IMG_3092.heic.xmp -all:all IMG_3092.heic
    1 image files updated
```

```
exiftool -tagsfromfile IMG_3092.heic.xmp -all:all IMG_3092.heic -v3
Setting new values from IMG_3092.heic.xmp
Sorry, XMP-x:XMPToolkit is unsafe for writing
Writing XMP-dc:Title
Writing XMP-digiKam:TagsList
Writing XMP-dc:Subject
Writing XMP-photoshop:SidecarForExtension
Sorry, ExifTool:ProcessingTime doesn't exist or isn't writable
Sorry, ExifTool:Now doesn't exist or isn't writable
Sorry, ExifTool:NewGUID doesn't exist or isn't writable
Writing XMP-xmp:ModifyDate
Sorry, File:MIMEType doesn't exist or isn't writable
Sorry, Composite:GPSPosition doesn't exist or isn't writable
Sorry, Composite:GPSLongitudeRef doesn't exist or isn't writable
Writing XMP-exif:GPSLongitude
Sorry, Composite:GPSLatitudeRef doesn't exist or isn't writable
Writing XMP-exif:GPSLatitude
Sorry, File:FileUserID is unsafe for writing
Sorry, File:FileTypeExtension doesn't exist or isn't writable
Sorry, File:FileType doesn't exist or isn't writable
Sorry, System:FileSize doesn't exist or isn't writable
Sorry, ExifTool:FileSequence doesn't exist or isn't writable
Sorry, File:FilePermissions is unsafe for writing
Sorry, System:FilePath doesn't exist or isn't writable
Sorry, File:FileName is unsafe for writing
Sorry, File:FileModifyDate is unsafe for writing
Sorry, System:FileInodeNumber doesn't exist or isn't writable
Sorry, System:FileInodeChangeDate doesn't exist or isn't writable
Sorry, System:FileHardLinks doesn't exist or isn't writable
Sorry, File:FileGroupID is unsafe for writing
Sorry, System:FileDeviceNumber doesn't exist or isn't writable
Sorry, System:FileDeviceID doesn't exist or isn't writable
Sorry, System:FileBlockSize doesn't exist or isn't writable
Sorry, System:FileBlockCount doesn't exist or isn't writable
Sorry, System:FileAttributes doesn't exist or isn't writable
Sorry, System:FileAccessDate doesn't exist or isn't writable
Sorry, ExifTool:ExifToolVersion doesn't exist or isn't writable
Sorry, File:Directory is unsafe for writing
Writing XMP-dc:Description
Writing XMP-photoshop:DateCreated
Writing XMP-xmp:CreateDate
======== IMG_3092.heic
Rewriting IMG_3092.heic...
  FileType = HEIC
  FileTypeExtension = HEIC
  MIMEType = video/quicktime
  Editing tags in: ItemInformation MOV Meta XMP
  Creating tags in: ItemInformation MOV Meta XMP
  Rewriting Meta
  Rewriting DataInfo
  Rewriting DataRef
  Rewriting ItemInformation
  Rewriting ItemReference
  Rewriting ItemProperties
  Rewriting ItemPropertyContainer
  Rewriting XMP
    - XMP-xmp:CreateDate = '2020-09-19T14:36:26'
    + XMP-xmp:CreateDate = '2020-09-19T14:36:26'
    - XMP-photoshop:DateCreated = '2020-09-19T14:36:26.719000-07:00'
    + XMP-photoshop:DateCreated = '2020-09-19T14:36:26.719000-07:00'
    - XMP-dc:Description-x-default = ''
    + XMP-dc:Description = ''
    - XMP-exif:GPSLatitude = '41,15.39396N'
    + XMP-exif:GPSLatitude = '41,15.39396N'
    - XMP-exif:GPSLongitude = '95,56.41542W'
    + XMP-exif:GPSLongitude = '95,56.41542W'
    - XMP-xmp:ModifyDate = '2020-09-19T14:36:26'
    + XMP-xmp:ModifyDate = '2020-09-19T14:36:26'
    - XMP-photoshop:SidecarForExtension = 'heic'
    + XMP-photoshop:SidecarForExtension = 'heic'
    - XMP-dc:Subject = 'foo/bar'
    + XMP-dc:Subject = 'foo/bar'
    - XMP-digiKam:TagsList = 'foo/bar'
    + XMP-digiKam:TagsList = 'foo/bar'
    - XMP-dc:Title-x-default = ''
    + XMP-dc:Title = ''
    1 image files updated
```

**Comment by narensankar0529** [comment]
Warning: [minor] Fixed incorrect list type for XMP-dc:Description - t2.xmp

**Comment by narensankar0529** [comment]
And I found by some searching online that this error would mean that the file XMP data won't be read by Windows Explorer as an example.

**Comment by narensankar0529** [comment]
So I figured out that exiftool does not like empty lists -

dc:description</dc:description>
dc:title</dc:title>

So if you don't output these at all when description and title are empty then there is no error. In your case you have valid entries for description and title so you don't get the error. Currently osxphotos outputs an empty list which exiftool does not like. So just deleting these two lines, then exiftool handles it correctly.

**Comment by RhetTbull** [comment]
It's easy to omit the fields if they're empty. However this could potentially lead to "incorrect" results: for example, user imports a photo that has title or subject in exif, deletes them in Photos, then exports. What's the correct behavior in this case? If omitted in XMP, the resulting Image file will still have the original (deleted) data. But if this causes issues for exiftool or Windows I guess the most correct thing is to omit the empty fields.

**Comment by narensankar0529** [comment]
So this is an issue currently anyway. Seems strange that Photos doesn't modify the exif data when Title and Description are deleted. So I just did the test as you describe and when exported via osxphotos - the exif in the file still has the Title and Description whereas the XMP has null fields. If the desired logic is then to use exiftool to write back to the file and delete the exif fields, the correct way is to omit the fields since otherwise exiftool will issue the warning anyway. 

So either way this is an issue. Just that if you omit the fields then exiftool won't issue a warning and if you write back to the file to update the exif Windows Explorer will correctly show empty fields

**Comment by RhetTbull** [comment]
> Seems strange that Photos doesn't modify the exif data when Title and Description are deleted.

Photos *never* modifies the original file.  All modifications happen only to edited images.  In the case of metadata, modifications only happen on export or added to the XMP when selecting "Export unmodified original"

If exporting via osxphotos the metadata in the exported file is never modified unless you use `--exiftool` option.

> If the desired logic is then to use exiftool to write back to the file and delete the exif fields, the correct way is to omit the fields since otherwise exiftool will issue the warning anyway.

In my testing, omitting the field does not cause exiftool to delete the metadata, it simply ignores any existing metadata.  Only including a null field causes exiftool to delete the metadata.

Here's some test data:

Original image with a title:

```
exiftool -title test.heic
Title                           : My Title
```

Import into Photos:
Photos shows title as : "My Title"

Delete the title in Photos and export with Photos, title in exif is unchanged
```
exiftool -title test.heic
Title                           : My Title
```

Apply the xmp sidecar that Photos creates, title from exif is not deleted:

```
exiftool -tagsfromfile test.xmp -all:all test.heic
    1 image files updated

exiftool -title test.heic
Title                           : My Title
```

export with osxphotos, still has title as per original exif
```
exiftool -title test.heic
Title                           : My Title
```

export with osxphotos xmp sidecar, use exiftool to apply the sidecar and title is deleted (null title in XMP)
```
osxphotos export ~/Desktop/export --uuid C48B5596-647F-4D9D-A8AB-94D676A93688 -V --sidecar xmp

exiftool -title test.heic
Title                           : My Title

exiftool -tagsfromfile test.heic.xmp -all:all test.heic
    1 image files updated

exiftool -title test.heic
Title                           :
```

export with osxphotos and --exiftool, title is *not* deleted -- this is unexpected and, from my perspective, a bug in osxphotos. I would have expected --exiftool to delete the title to match what's in Photos so I'll need to look at this and determine if that's still the best behavior:

```
osxphotos export ~/Desktop/export --uuid C48B5596-647F-4D9D-A8AB-94D676A93688 -V --exiftool

exiftool -title test.heic
Title                           : My Title
```

Also, I never get any warnings when writing xmp with `exiftool -tagsfromfile`

**Comment by RhetTbull** [comment]
What exiftool command are you using that causes the warning?

**Comment by RhetTbull** [comment]
I created a test case to export a several files with XMP sidecars (all with various combinations of metadata) and then try to apply to the tags from XMP with exiftool but I've not been able to get exiftool to generate warnings for any of the XMP files.  If you post the exact command line you're using with exiftool I'll add it to the test case so I can then work backwards to eliminate the warnings.

**Comment by narensankar0529** [comment]
exiftool -tagsfromfile t2.xmp -all:all t1.xmp

I am trying to merge the xmp generated by osxphotos with xmp data modified externally. It works correctly without any issues in that the final output is correct - except for the warning that exiftool generates.

**Comment by RhetTbull** [comment]
OK, I was applying the xmp tags to jpeg, not merging.  I'm able to replicate this now.  I'll update the test case and work on fixing these warnings.

**Comment by RhetTbull** [comment]
Interesting -- this appears to be a problem with both source and destination XMP file.  When I merge XMP files as you did above with both source and destination XMPs generated by osxphotos I get the warning:
`Warning: [minor] Fixed incorrect list type for XMP-dc:Description - test.xmp`

However, when I then merge again the osxphotos generated XMP with an XMP generated by exiftool, I don't get the warning.  So exiftool is happy if writing to one of it's own XMP but not happy with the externally generated XMP you have.  I will still work to fix osxphotos to conform to exiftool format.

**Comment by RhetTbull** [comment]
I think this is fixed in [v0.39.18](https://github.com/RhetTbull/osxphotos/releases/tag/v0.39.18).  I've fixed XMP sidecar output to match exiftool as much as possible.  Let me know if you still have warnings.

**Comment by narensankar0529** [comment]
Works great! Thanks a lot.
